pkg/platform/kube/monitoring/function.go

```diff
@@ -0,0 +1,217 @@
+	}
+
+	newFunctionMonitor.logger.DebugWith("Created function monitor",
+		"namespace", namespace,
+		"pollingInterval", pollingInterval)
+
+	return newFunctionMonitor, nil
+}
+
+func (fm *FunctionMonitor) Start() error {
+	fm.logger.InfoWith("Starting",
+		"namespace", fm.namespace)
+
+	// create stop channel
+	fm.stopChan = make(chan struct{}, 1)
+
+	// spawn a goroutine for function monitoring
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				callStack := debug.Stack()
+				fm.logger.ErrorWith("Panic caught while starting function monitoring",
+					"err", err,
+					"stack", string(callStack))
+			}
+		}()
+		for {
+			select {
+			case <-time.After(fm.pollingInterval):
+				if err := fm.checkFunctionStatuses(); err != nil {
+					fm.logger.WarnWith("Failed check function statuses",
+						"namespace", fm.namespace,
+						"err", errors.Cause(err))
+				}
+
+			case <-fm.stopChan:
+				fm.logger.DebugWith("Stopped function monitoring",
+					"namespace", fm.namespace)
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (fm *FunctionMonitor) Stop() {
+	fm.logger.InfoWith("Stopping function monitoring", "namespace", fm.namespace)
+
+	// post to channel
+	if fm.stopChan != nil {
+		fm.stopChan <- struct{}{}
+	}
+}
+
+func (fm *FunctionMonitor) checkFunctionStatuses() error {
+	functions, err := fm.nuclioClientSet.NuclioV1beta1().NuclioFunctions(fm.namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Failed to list functions")
+	}
+
+	var errGroup errgroup.Group
+	for _, function := range functions.Items {
+		function := function
+		errGroup.Go(func() error {
+			return fm.checkFunctionStatus(&function)
+		})
+	}
+	return errGroup.Wait()
+}
+
+func (fm *FunctionMonitor) checkFunctionStatus(function *nuclioio.NuclioFunction) error {
+
+	if fm.shouldSkipCheckFunctionStatus(function) {
+
+		// skip check for function status
+		return nil
+	}
+
+	functionDeployment, err := fm.kubeClientSet.
+		AppsV1().
+		Deployments(function.Namespace).
+		Get(kube.DeploymentNameFromFunctionName(function.Name), metav1.GetOptions{})
+	if err != nil {
+		fm.logger.WarnWith("Failed to get function deployment",
+			"functionName", function.Name,
+			"functionNamespace", function.Namespace)
+		return nil
+	}
+
+	stateChanged := false
+	functionIsAvailable := fm.isAvailable(functionDeployment)
+	if functionIsAvailable && function.Status.State == functionconfig.FunctionStateError {
+		fm.logger.InfoWith("Function has recovered, removing error and set state as ready",
+			"functionName", function.Name,
+			"functionNamespace", function.Namespace,
+			"functionStatusMessage", function.Status.Message)
+		function.Status.State = functionconfig.FunctionStateReady
+		function.Status.Message = ""
+		stateChanged = true
+	} else if !functionIsAvailable && function.Status.State == functionconfig.FunctionStateReady {
+		fm.logger.InfoWith("Function has become unhealthy, setting error state",
```

Comment:
This and the above logs are redundant since you're logging every update below (you can just make it info level)